ambari-server/src/main/java/org/apache/ambari/server/serveraction/upgrades/ConfigureAction.java

```diff
@@ -717,66 +724,65 @@ private static String mask(Masked mask, String value) {
-          configType, targetPropertyKey, ifKey, ifType);
+          configType, targetPropertyKey, operationItem.ifKey, operationItem.ifType);
         isAllowed = false;
       }
       // if-key/if-type and no value - set only is present
-    } else if (ifKeyIsNotBlank && ifTypeIsNotBlank && ifValueIsBlank && ifKeyState == PropertyKeyState.PRESENT) {
-      boolean keyPresent = getDesiredConfigurationKeyPresence(cluster, ifType, ifKey);
+    } else if (ifKeyIsNotBlank && ifTypeIsNotBlank && ifValueIsBlank
+      && operationItem.ifKeyState == PropertyKeyState.PRESENT) {
+
+      boolean keyPresent = getDesiredConfigurationKeyPresence(cluster, operationItem.ifType, operationItem.ifKey);
       if (!keyPresent) {
         LOG.info("Skipping property operation for {}/{} as the key {} for {} is not present",
-          configType, targetPropertyKey, ifKey, ifType);
+          configType, targetPropertyKey, operationItem.ifKey, operationItem.ifType);
         isAllowed = false;
       }
       // if-key/if-type and a value to check - set only if values match
     } else if (ifKeyIsNotBlank && ifTypeIsNotBlank && !ifValueIsBlank) {
-      String ifConfigType = ifType;
-      String checkValue = getDesiredConfigurationValue(cluster, ifConfigType, ifKey);
+      String checkValue = getDesiredConfigurationValue(cluster, operationItem.ifType, operationItem.ifKey);
 
       // the check value is blank and there is an if-key-state of ABSENT - in
       // this case, it means set the value if it matches or if it's absent
-      if (ifKeyState == PropertyKeyState.ABSENT) {
-        boolean keyPresent = getDesiredConfigurationKeyPresence(cluster, ifType, ifKey);
+      if (operationItem.ifKeyState == PropertyKeyState.ABSENT) {
+        boolean keyPresent = getDesiredConfigurationKeyPresence(cluster, operationItem.ifType, operationItem.ifKey);
         if (!keyPresent) {
           return true;
         }
       }
 
       // the if-key was found, so we need to do a comparison
-      if (!StringUtils.equalsIgnoreCase(ifValue, checkValue)) {
-        // skip adding
-        LOG.info("Skipping property operation for {}/{} as the value {} for {}/{} is not equal to {}",
-                 configType, targetPropertyKey, checkValue, ifConfigType, ifKey, ifValue);
-        isAllowed = false;
+      if (operationItem.ifValueMatchType == IfValueMatchType.PARTIAL) {
+        if (!StringUtils.containsIgnoreCase(checkValue, operationItem.ifValue) ^ operationItem.ifValueNotMatched) {
+          // skip adding
+          LOG.info("Skipping property operation for {}/{} as the value {} for {}/{} is not equal to {}",
+            configType, targetPropertyKey, checkValue, operationItem.ifType, operationItem.ifKey, operationItem.ifValue);
+          isAllowed = false;
+        }
+      } else {
+        if (!StringUtils.equalsIgnoreCase(operationItem.ifValue, checkValue) ^ operationItem.ifValueNotMatched) {
+          // skip adding
```

Comment:
if body seems to be just equal to the body above, including log messsage. How about uniting them and passing a parameter?

Reply:
@dlysnichenko it is not identical as i have forgot to change info message....